pkg/controlplane/controller/leaderelection/leaderelection_controller.go

```diff
@@ -0,0 +1,633 @@
+		AddFunc: func(obj interface{}) {
+			c.enqueueLease(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.enqueueLease(newObj)
+		},
+		DeleteFunc: func(oldObj interface{}) {
+			c.leaseDeleted(oldObj)
+		},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	leaseCandidateSynced, err := leaseCandidateInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueCandidate(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.enqueueCandidate(newObj)
+		},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	c.leaseSynced = leaseSynced.HasSynced
+	c.leaseCandidateSynced = leaseCandidateSynced.HasSynced
+	return c, nil
+}
+
+func (c *Controller) enqueueLease(obj any) {
+	if lease, ok := obj.(*v1.Lease); ok {
+		// TODO: handle namespaces
+		key, err := controller.KeyFunc(lease)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("cannot get name of object %v: %w", lease, err))
+		}
+		c.leaseQueue.Add(key)
+	}
+}
+
+func (c *Controller) enqueueCandidate(obj any) {
+	if lease, ok := obj.(*v1alpha1.LeaseCandidate); ok {
+		// TODO: handle namespaces
+		// key, err := controller.KeyFunc(lease)
+		// if err != nil {
+		// 	utilruntime.HandleError(fmt.Errorf("cannot get name of object %v: %w", lease, err))
+		// }
+		_ = c.reconcileLeaseCandidate(lease)
+		// How to handle invalid lease?
+	}
+}
+
+func (c *Controller) leaseDeleted(oldObj any) {
+	klog.Infof("leaseDeleted")
+	if lease, ok := oldObj.(*v1.Lease); ok {
+		// TODO: Check if this is a lease that needs a leader election? We may need to maintain
+		// a set of leader election leases?  For prototype purposes can we just scan the leases?
+		err := c.scheduleElection(leaderLeaseID{namespace: lease.Namespace, name: lease.Name})
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("cannot start election for %v: %w", lease, err))
+		}
+	}
+}
+
+func (c *Controller) runWorker(ctx context.Context) {
+	klog.Infof("runWorker")
+	go c.runElectionLoop(ctx.Done())
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	key, shutdown := c.leaseQueue.Get()
+	if shutdown {
+		return false
+	}
+	defer c.leaseQueue.Done(key)
+
+	err := func() error {
+		return c.reconcile(ctx, key)
+	}()
+
+	if err == nil {
+		c.leaseQueue.AddAfter(key, 10*time.Second)
+		return true
+	}
+
+	utilruntime.HandleError(err)
+	klog.Infof("processNextWorkItem.AddRateLimited: %v", key)
+	c.leaseQueue.AddRateLimited(key)
+
+	return true
+}
+
+func (c *Controller) runElectionLoop(stopCh <-chan struct{}) {
+	klog.Infof("runElectionLoop")
+	ctx := context.Background()
+	for {
+		select {
+		case <-stopCh:
+			return
+		case e := <-c.electionCh:
+			select {
+			case <-stopCh:
+				return
+			// TODO; Isn't this always true? Is this for dealing with time drift?
+			case <-time.After(time.Until(e.electionStart)): // elections are time ordered in the channel
```

Comment:
you may want to use a ticker and stop the timer just as Jordan did here https://github.com/moby/spdystream/pull/101/files

Reply:
not needed now.